iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relocatable_ptr.inl

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_RELOCATABLE_POINTER_RELOCATABLE_PTR_INL
+#define IOX_HOOFS_RELOCATABLE_POINTER_RELOCATABLE_PTR_INL
+
+#include "iceoryx_hoofs/internal/relocatable_pointer/relocatable_ptr.hpp"
+
+namespace iox
+{
+namespace rp
+{
+template <typename T>
+relocatable_ptr<T>::relocatable_ptr(T* ptr)
+{
+    m_offset = to_offset(ptr);
+}
+
+template <typename T>
+relocatable_ptr<T>::relocatable_ptr(const relocatable_ptr& other)
+{
+    m_offset = to_offset(other.get());
+}
+
+template <typename T>
+relocatable_ptr<T>::relocatable_ptr(relocatable_ptr&& other)
+{
+    m_offset = to_offset(other.get());
+    other.m_offset = NULL_POINTER_OFFSET;
+}
+
+template <typename T>
+relocatable_ptr<T>& relocatable_ptr<T>::operator=(const relocatable_ptr& rhs)
+{
+    if (this != &rhs)
+    {
+        m_offset = to_offset(rhs.get());
+    }
+    return *this;
+}
+
+template <typename T>
+relocatable_ptr<T>& relocatable_ptr<T>::operator=(relocatable_ptr&& rhs)
+{
+    if (this != &rhs)
+    {
+        m_offset = to_offset(rhs.get());
+        rhs.m_offset = NULL_POINTER_OFFSET;
+    }
+    return *this;
+}
+
+template <typename T>
+T* relocatable_ptr<T>::get()
+{
+    return from_offset(m_offset);
+}
+
+template <typename T>
+const T* relocatable_ptr<T>::get() const
+{
+    return from_offset(m_offset);
+}
+
+template <typename T>
+template <typename S>
+S& relocatable_ptr<T>::operator*()
+{
+    // not actually evaluated in the error case (compiler fails earlier since S = void leads to void&)
+    static_assert(!std::is_same<S, void>::value, "relocatable_ptr<void> does not support operator*");
```

Comment:
do we really need a second template parameter for this check?

Reply:
Yes, otherwise the class template instantiation will fail at class scope for `T=void` regardless of whether we use `operator*`, which is not what we want. By introducing S compilation will only fail if we actually try to use it. There may other ways to accomplish that, but to my knowledge they are all similarly relying on postponing the compilation of the operator by using an additional template argument.

They are also not shorter. E.g. `enable_if`is overkill here since we do not switch between alternatives (we just fail in the void case).